storage/test/sparse_merkle_storage/db_adapter_property_test.cpp

```diff
@@ -0,0 +1,313 @@
+    if (it != std::cend(updates)) {
+      ret = std::make_pair(BlockId{i}, it->second);
+    }
+  }
+  return ret;
+}
+
+struct IDbAdapterTest {
+  virtual std::shared_ptr<IDBClient> db() const = 0;
+  virtual std::string type() const = 0;
+  virtual bool enforceMultiVersionedKeys() const = 0;
+  virtual ~IDbAdapterTest() noexcept = default;
+};
+
+template <typename Database, bool multiVersionedKeys = false>
+struct DbAdapterTest : public IDbAdapterTest {
+  std::shared_ptr<IDBClient> db() const override { return Database::create(); }
+  std::string type() const override {
+    return Database::type() + (enforceMultiVersionedKeys() ? "_enforcedMultiVerKey" : "_noMultiVerKeyEnforced");
+  }
+  bool enforceMultiVersionedKeys() const override { return multiVersionedKeys; }
+};
+
+using db_adapter_block_tests = ParametrizedTest<std::shared_ptr<IDbAdapterTest>>;
+using db_adapter_kv_tests = ParametrizedTest<std::shared_ptr<IDbAdapterTest>>;
+
+// Test that fetched block parameters match the ones that were used to create the block.
+RC_GTEST_PROP(block,
+              create,
+              (const SetOfKeyValuePairs &updates, const ParentDigest &parentDigest, const Hash &stateHash)) {
+  const auto block = create(updates, parentDigest.data(), stateHash);
+  RC_ASSERT(getData(block) == updates);
+  RC_ASSERT(equal(getParentDigest(block), parentDigest));
+  RC_ASSERT(getStateHash(block) == stateHash);
+}
+
+// Test that the blockchain property is maintained. In pseudocode:
+//  * [getData(block) == blockUpdate] - the passed updates are the same as the fetched ones after adding the block
+//  * [getParentDigest(block) == blockDigest(prevBlock)] - blockchain property
+//  * [getLastReachableBlock() == getLatestBlock() == blockUpdates.size()] - last reachable block is equal to the latest
+//  know block ID
+TEST_P(db_adapter_block_tests, blockchain_property) {
+  const auto test = [this](const std::vector<SetOfKeyValuePairs> &blockUpdates) {
+    RC_PRE(!hasEmptyValue(blockUpdates));
+
+    auto adapter = DBAdapter{GetParam()->db()};
+    RC_ASSERT(addBlocks(blockUpdates, adapter));
+
+    RC_ASSERT(adapter.getLastReachableBlock() == blockUpdates.size());
+    RC_ASSERT(adapter.getLatestBlock() == blockUpdates.size());
+
+    for (auto i = 0ul; i < blockUpdates.size(); ++i) {
+      auto block = Sliver{};
+      RC_ASSERT(adapter.getBlockById(i + 1, block).isOK());
+      RC_ASSERT(getData(block) == blockUpdates[i]);
+      if (i == 0) {
+        RC_ASSERT(equal(getParentDigest(block), zeroDigest));
+      } else {
+        auto prevBlock = Sliver{};
+        RC_ASSERT(adapter.getBlockById(i, prevBlock).isOK());
+        RC_ASSERT(equal(getParentDigest(block), blockDigest(i, prevBlock)));
+      }
+    }
+  };
+
+  ASSERT_TRUE(rc::check(test));
+}
+
+// Test that when adapter.addBlock(block, getLastReachableBlock() + 1) is called and there is a hole between the
+// blockchain and the state transfer chain, the block is added at the back of the blockchain.
```

Comment:
Maybe clarify to say after `LastReachableBlock` (in the gap).

Reply:
Done. Also changed the test so that we explicitly have a gap of at least 2 blocks. That way, we will expect that the chains are not linked at the end of the test. Linking is tested in the state transfer simulation test. Another test we might want is a test that adds blocks in random order and verifies that linking works.